connectors/sources/notion.py

```diff
@@ -96,6 +96,48 @@ async def get_via_session(self, url):
             else:
                 raise
 
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def fetch_results(
+        self, function: Callable[..., Awaitable[Any]], next_cursor=None, **kwargs: Any
+    ):
+        try:
+            return await function(start_cursor=next_cursor, **kwargs)
+        except Exception as exception:
+            if exception.code == "rate_limited" and exception.status == 429:
+                retry_after = (
+                    exception.headers.get("retry-after") or DEFAULT_RETRY_SECONDS
+                )
+                self._logger.info(
+                    f"Connector will attempt to retry after {int(retry_after)} seconds."
+                )
+                await self._sleeps.sleep(int(retry_after))
+                msg = "Rate limit exceeded."
+                raise Exception(msg) from exception
+            else:
+                raise
+
+    async def async_iterate_paginated_api(
+        self, function: Callable[..., Awaitable[Any]], **kwargs: Any
+    ):
+        """Return an async iterator over the results of any paginated Notion API."""
+        next_cursor = kwargs.pop("start_cursor", None)
+        while True:
+            try:
+                response = await self.fetch_results(function, next_cursor, **kwargs)
+                if response:
+                    for result in response.get("results"):
+                        yield result
+                    next_cursor = response.get("next_cursor")
+                    if not response["has_more"] or next_cursor is None:
+                        return
+            except Exception as error:
+                raise error
```

Comment:
This seems entirely redundant. Unless you're going to handle `error` or log something, why bother catching it just to re-raise?